### PR TITLE
refactor(executor): refactor CopyFromFile using DataChunkBuilder

### LINKED ISSUE #618: executor: refactor CopyFromFile using DataChunkBuilder

### PR DESCRIPTION
**Pull request by noneback**
refactor CopyFromFile using DataChunkBuilder

close #618

### CODE REVIEW

File: src/executor/copy_from_file.rs [hunk 1]
```diff
@@ -78,14 +79,9 @@ impl CopyFromFileExecutor {
...
-                .column_types()
-                .iter()
-                .map(|ty| ArrayBuilderImpl::with_capacity(PROCESSING_WINDOW_SIZE, ty))
-                .collect_vec();
+            // create chunk builders
+            let mut chunk_builder =
+                DataChunkBuilder::new(self.plan.logical().column_types(), PROCESSING_WINDOW_SIZE);
 
             // read records and push to array builder
             for _ in 0..PROCESSING_WINDOW_SIZE {
```
Comment (xxchan) [inline src/executor/copy_from_file.rs:87]: I think `chunk_builder` can be created outside `while !finished` loop, and `for _ in 0..PROCESSING_WINDOW_SIZE` loop is not needed any more?

Comment (skyzh) [reply src/executor/copy_from_file.rs:87]: +1

File: src/array/data_chunk_builder.rs [hunk 2]
```diff
@@ -57,6 +57,31 @@ impl DataChunkBuilder {
         }
     }
 
+    /// Push a row of str in the Iterator.
+    ///
+    /// The row is accepted as an iterator of &str, and it's required that the size of row
+    /// should be the same as the number of columns.
+    ///
+    /// A [`DataChunk`] will be returned while `size == capacity`, and it should always be handled
+    /// correctly.
```
Comment (xxchan) [inline src/array/data_chunk_builder.rs:66]:
```suggestion
    /// Push a row of `str` in the Iterator.
    ///
    /// The row is accepted as an iterator of `&str`, and it's required that the size of row
    /// should be the same as the number of columns.
    ///
    /// A [`DataChunk`] will be returned while `size == capacity`, and it should always be handled
    /// correctly.
    #[must_use]
```

File: src/array/data_chunk_builder.rs [hunk 3]
```diff
@@ -57,6 +57,31 @@ impl DataChunkBuilder {
...
+    pub fn push_str_row<'a>(
+        &mut self,
+        row: impl IntoIterator<Item = &'a str>,
+    ) -> Result<Option<DataChunk>, ConvertError> {
+        for (builder, r) in self.array_builders.iter_mut().zip_eq(row) {
+            if let Some(e) = builder.push_str(r).err() {
+                return Err(e);
+            }
```
Comment (xxchan) [inline src/array/data_chunk_builder.rs:74]: A style suggestion:
```suggestion
            builder.push_str(r)?;
```

File: src/executor/copy_from_file.rs (hunk 1 above)
Comment (noneback) [reply src/executor/copy_from_file.rs:87]: Ok, I will try to fix it. I am not so familiar with rust. 
Sorry for that.

File: src/executor/copy_from_file.rs [hunk 4]
```diff
@@ -104,27 +100,33 @@ impl CopyFromFileExecutor {
...
 
             // send data chunk
-            let chunk: DataChunk = array_builders.into_iter().collect();
-
             #[allow(clippy::collapsible_if)]
-            if chunk.cardinality() > 0 {
-                if token.is_cancelled() || tx.blocking_send(chunk).is_err() {
-                    return Err(ExecutorError::Abort);
+            if let Some(chunk) = { chunk_builder }.take() {
+                if chunk.cardinality() > 0 {
```
Comment (xxchan) [inline src/executor/copy_from_file.rs:126]: From the doc of `DataChunkBuilder::take`: If there are no remaining rows, `None` will be returned.

So I guess `if chunk.cardinality() > 0 {` can be removed. (And maybe also `#[allow(clippy::collapsible_if)]`?)

File: src/executor/copy_from_file.rs (hunk 1 above)
Comment (xxchan) [reply src/executor/copy_from_file.rs:87]: No need to be sorry. Just try it! 😄 

You may also refer to other refactors e.g. https://github.com/risinglightdb/risinglight/pull/623/files to see how the pattern changes

Comment (noneback) [reply src/executor/copy_from_file.rs:87]: Thx for your patience! It's of great help:)

**Review by xxchan — COMMENTED** [review]
Btw the tests failed. I haven't taken a look yet. You may try to fix it

File: src/array/data_chunk_builder.rs (hunk 2 above)
Comment (noneback) [reply src/array/data_chunk_builder.rs]: `Result` has already been marked as `#[must_use]`, so I guess there is no need to add one.

File: src/executor/copy_from_file.rs [hunk 5]
```diff
@@ -75,59 +76,58 @@ impl CopyFromFileExecutor {
...
+                        }
+                        Ok(v)
+                    })
+                    .collect();
+
+            // push a raw str row
+            let chunk = match chunk_builder.push_str_row(str_row_data?) {
+                Ok(chunk) => chunk,
+                Err(e) => return Err(ExecutorError::Convert(e)),
+            };
```
Comment (xxchan) [inline src/executor/copy_from_file.rs:112]: Can be simplified and merged below
```suggestion
```

File: src/executor/copy_from_file.rs [hunk 6]
```diff
@@ -75,59 +76,58 @@ impl CopyFromFileExecutor {
...
 
-            #[allow(clippy::collapsible_if)]
-            if chunk.cardinality() > 0 {
+            // update progress bar
+            size_count += record.as_slice().as_bytes().len();
+            bar.set_position(size_count as u64);
+
+            // send chunk if necessary
+            if let Some(chunk) = chunk {
```
Comment (xxchan) [inline src/executor/copy_from_file.rs:119]:
```suggestion
            if let Some(chunk) = chunk_builder.push_str_row(str_row_data?)?; {
```

**Review by xxchan — APPROVED** [review]
Others LGTM!

File: src/executor/copy_from_file.rs [hunk 7]
```diff
@@ -75,59 +76,52 @@ impl CopyFromFileExecutor {
...
+                        }
+                        Ok(v)
+                    })
+                    .collect();
 
-            #[allow(clippy::collapsible_if)]
-            if chunk.cardinality() > 0 {
+            // update progress bar
+            size_count += record.as_slice().as_bytes().len();
+            bar.set_position(size_count as u64);
```
Comment (skyzh) [inline src/executor/copy_from_file.rs:110]: update bar per-record will have significant performance impact. Consider only updating it when a chunk is produced.

**Review by skyzh — COMMENTED** [review]
Rest LGTM, good work!

**Review by skyzh — APPROVED** [review]
Thanks!

**Pull request merged**